File: nodejs/caesar.test.mjs
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

const upper = [...'ABCDEFGHIJKLMNOPQRSTUVWXYZÆØÅ'];
const lower = [...'abcdefghijklmnopqrstuvwxyzæøå'];
const size  = upper.length;

const idx = new Map([
  ...upper.map((ch, i) => [ch, i]),
  ...lower.map((ch, i) => [ch, ~i]),
]);

function cipher(text, shift) {
  shift = ((shift % size) + size) % size;
  return [...text].map(ch => {
    const pos = idx.get(ch);
    if (pos === undefined) return ch;
    return pos >= 0
      ? upper[(pos + shift) % size]
      : lower[(~pos + shift) % size];
  }).join('');
}

describe('encrypt', () => {
  const cases = [
    ['ABC',            'BCD',            1],
    ['abc',            'bcd',            1],
    ['XYZ',            'YZÆ',            1],
    ['ÆØÅ',            'ØÅA',            1],
    ['Hello, World!',  'Mjqqt, Øtwqi!',  5],
    ['Hei æøå!',       'Mjn cde!',       5],
    ['Hello Æøå!',     'Khoor Abc!',     3],
    ['ABC',            'ABC',            0],
    ['ABC',            'ABC',            29],
    ['123 !?',         '123 !?',         7],
  ];
  for (const [input, want, shift] of cases) {
    it(`cipher(${JSON.stringify(input)}, ${shift}) === ${JSON.stringify(want)}`, () => {
      assert.equal(cipher(input, shift), want);
    });
  }
});

describe('decrypt', () => {
  const cases = [
    ['BCD',            'ABC',            1],
    ['bcd',            'abc',            1],
    ['YZÆ',            'XYZ',            1],
    ['ØÅA',            'ÆØÅ',            1],
    ['Mjqqt, Øtwqi!',  'Hello, World!',  5],
    ['ABC',            'ABC',            0],
    ['ABC',            'ABC',            29],
  ];
  for (const [input, want, shift] of cases) {
    it(`cipher(${JSON.stringify(input)}, -${shift}) === ${JSON.stringify(want)}`, () => {
      assert.equal(cipher(input, -shift), want);
    });
  }
});

describe('roundtrip', () => {
  const cases = [
    ['Hello, World!',                         5],
    ['Hei på deg, æøå!',                      13],
    ['ABCDEFGHIJKLMNOPQRSTUVWXYZÆØÅ',         7],
    ['abcdefghijklmnopqrstuvwxyzæøå',         7],
    ['The quick brown fox!',                  29],
  ];
  for (const [text, shift] of cases) {
    it(`roundtrip ${JSON.stringify(text)} shift=${shift}`, () => {
      assert.equal(cipher(cipher(text, shift), -shift), text);
    });
  }
});

File: java/CaesarTest.java
import java.util.*;

public class CaesarTest {

    record Case(String input, String want, int shift) {}

    static final List<Case> ENCRYPT = List.of(
        new Case("ABC",           "BCD",           1),
        new Case("abc",           "bcd",           1),
        new Case("XYZ",           "YZ\u00C6",      1),
        new Case("\u00C6\u00D8\u00C5", "\u00D8\u00C5A", 1),
        new Case("Hello, World!", "Mjqqt, \u00D8twqi!", 5),
        new Case("Hei \u00E6\u00F8\u00E5!", "Mjn cde!", 5),
        new Case("Hello \u00C6\u00F8\u00E5!", "Khoor Abc!", 3),
        new Case("ABC",   "ABC",   0),
        new Case("ABC",   "ABC",   29),
        new Case("123 !?", "123 !?", 7)
    );

    static final List<Case> DECRYPT = List.of(
        new Case("BCD",    "ABC",   1),
        new Case("bcd",    "abc",   1),
        new Case("YZ\u00C6", "XYZ", 1),
        new Case("\u00D8\u00C5A", "\u00C6\u00D8\u00C5", 1),
        new Case("Mjqqt, \u00D8twqi!", "Hello, World!", 5),
        new Case("ABC",   "ABC",   0),
        new Case("ABC",   "ABC",   29)
    );

    static final List<Case> ROUNDTRIP = List.of(
        new Case("Hello, World!",                         null, 5),
        new Case("Hei p\u00E5 deg, \u00E6\u00F8\u00E5!", null, 13),
        new Case("ABCDEFGHIJKLMNOPQRSTUVWXYZ\u00C6\u00D8\u00C5", null, 7),
        new Case("abcdefghijklmnopqrstuvwxyz\u00E6\u00F8\u00E5", null, 7),
        new Case("The quick brown fox!", null, 29)
    );

    public static void main(String[] args) {
        int pass = 0, fail = 0;

        for (Case c : ENCRYPT) {
            String got = Caesar.cipher(c.input(), c.shift());
            if (got.equals(c.want())) { pass++; }
            else { fail++; System.err.printf("FAIL encrypt(%s, %d): got %s, want %s%n", c.input(), c.shift(), got, c.want()); }
        }

        for (Case c : DECRYPT) {
            String got = Caesar.cipher(c.input(), -c.shift());
            if (got.equals(c.want())) { pass++; }
            else { fail++; System.err.printf("FAIL decrypt(%s, %d): got %s, want %s%n", c.input(), c.shift(), got, c.want()); }
        }

        for (Case c : ROUNDTRIP) {
            String enc = Caesar.cipher(c.input(), c.shift());
            String dec = Caesar.cipher(enc, -c.shift());
            if (dec.equals(c.input())) { pass++; }
            else { fail++; System.err.printf("FAIL roundtrip(%s, %d): got %s%n", c.input(), c.shift(), dec); }
        }

        System.out.printf("%d/%d tests passed%n", pass, pass + fail);
        if (fail > 0) System.exit(1);
    }
}
